src/dc_array.rs

```diff
@@ -1,204 +1,265 @@
 use crate::dc_location::dc_location;
 use crate::dc_tools::*;
 use crate::types::*;
-use crate::x::*;
-
-const DC_ARRAY_MAGIC: uint32_t = 0x000a11aa;
 
 /* * the structure behind dc_array_t */
-#[derive(Copy, Clone)]
-#[repr(C)]
-pub struct dc_array_t {
-    pub magic: uint32_t,
-    pub allocated: size_t,
-    pub count: size_t,
-    pub type_0: libc::c_int,
-    pub array: *mut uintptr_t,
+#[derive(Clone)]
+pub enum dc_array_t {
+    Locations(Vec<dc_location>),
+    Uint(Vec<uintptr_t>),
```

Comment:
Maybe call this variant `UintPtr`, to indicate that it has pointers, not regular uints?

Reply:
Nobody uses `dc_array_add_ptr` anymore, because only locations were stored this way.